src/Chainweb/NodeVersion.hs

```diff
@@ -0,0 +1,137 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- |
+-- Module: Chainweb.NodeVersion
+-- Copyright: Copyright © 2019 Kadena LLC.
+-- License: MIT
+-- Maintainer: Lars Kuhtz <lars@kadena.io>
+-- Stability: experimental
+--
+-- TODO
+--
+module Chainweb.NodeVersion
+( NodeVersion(..)
+, minAcceptedVersion
+, isAcceptedVersion
+, getNodeVersion
+) where
+
+import Control.DeepSeq
+-- import Control.Retry
+
+import Data.Aeson
+import Data.Bifunctor
+import qualified Data.CaseInsensitive as CI
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+
+import GHC.Generics
+
+import qualified Network.HTTP.Client as HTTP
+
+-- internal modules
+
+import Chainweb.HostAddress
+import Chainweb.RestAPI.Utils
+import Chainweb.Utils
+import Chainweb.Version
+
+-- -------------------------------------------------------------------------- --
+-- Chainweb Node Version
+
+newtype NodeVersion = NodeVersion { _getNodeVersion :: [Int] }
+    deriving (Show, Eq, Ord, Generic, NFData)
+
+instance HasTextRepresentation NodeVersion where
+    toText (NodeVersion l) = T.intercalate "." $ sshow <$> l
+    fromText = fmap NodeVersion . traverse treadM . T.splitOn "."
+    {-# INLINE toText #-}
+    {-# INLINE fromText #-}
```

Comment:
Consider instead [my versions library](http://hackage.haskell.org/package/versions) for parsing these.

Reply:
I think it's not needed in this case. We don't have to represent arbitrary version numbers but only chainweb-node version numbers that will always be a lexicographical list of `Int`. Taking a new dependency and using a more complex type seems overkill for that.

Reply:
I think, the versions library is great for representing heterogenous sets of version from different packages. But here we deal with just a single package and we have full control over the format.